lib/common/ipc_client.c

```diff
@@ -693,6 +695,11 @@ pcmk__send_ipc_request(pcmk_ipc_api_t *api, const xmlNode *request)
             rc = crm_ipc_read(api->ipc);
 
             if (rc == -EAGAIN) {
+                /* 'crmadmin --dc_lookup' would stuck if called when the dc is offine.
+                 * so let's give it 5 seconds.
+                 */
+                if (timeout < time(NULL))
+                    break;
```

Comment:
crmadmin already has a `--timeout` option that it passes to `pcmk__designated_controller()`. If a user passes the option, it takes effect via that function setting the dispatch to polling and calling `poll_until_reply()`. Prior to a6ec43e3, the timeout defaulted to 30s, but as of that commit the default became sync (blocking). That was probably a bad idea.

The simplest fix would be to restore the default `--timeout` to 30s in crmadmin.c. That would mean there would no longer be a way to specify sync (blocking) from crmadmin. With a little more effort we could distinguish between the user not specifying `--timeout` (defaulting to 30s) and the user explicitly specifying `--timeout 0` for blocking, but that's probably not important to anyone.

Separately, the tight loop here calling `crm_ipc_read()` was always a poor design. `crm_ipc_read()` calls `qb_ipcc_event_recv()`, which has a timeout argument. However it always passes 0 (nonblocking). Instead of the loop, we should pass whatever timeout was requested. Since `crm_ipc_read()` is public API, we'd have to create a new equivalent function with a timeout argument and make `crm_ipc_read()` a wrapper for it that always passes 0 timeout. That's not needed for this PR though.

Reply:
Thank you for the review. I have completely redone the change.